M3 183/concurrent variant calling

doron-st opened the pull request:
1. run variant recalling with multiprocess (partition intervals_list, and run process per sublist)
2. accept either fastq pairs or analysis_ready bam files in a single workflow - a mix is also possible

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -326,41 +341,92 @@ process applyBQSR {
-            --extension 30 \\
-            --tmp-dir \$PWD
+    """
+    #!/usr/bin/python
```
maybe we should move it to a separate py file 
This file beggings to get a bit too long
WDYT?

doron-st replied:
Done

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -326,41 +341,92 @@ process applyBQSR {
-            --tmp-dir \$PWD
+    """
+    #!/usr/bin/python
+    import glob, os, subprocess
```
what about these imports? nextflow ensures them to be available?
if not, then need to remember to add them (by docker?)

doron-st replied:
nextflow doesn't install anything for us. 
Its all in the responsibility of the docker container (in case we use one).
I think in this case, these are standard python packages, and if not, they are already included in the container we use for this "process"

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -326,41 +341,92 @@ process applyBQSR {
+    """
+    set -eu
+    ls -1 | grep "vcf.gz\$" > inputs.list
+    java -Xms2000m -jar $gotc_path/picard.jar MergeVcfs \\
```
the Xmx is not needed here?

doron-st replied:
Needed - done

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -326,41 +341,92 @@ process applyBQSR {
 }
 
-//Step 10 - recall ref-panel variants from analysis ready bam file
+// Step 10/1 -  Recall ref-panel variants from analysis ready bam files (given, or generated)
```
what's the meaning of the /1 here?

doron-st replied:
In case a bam input is given, this will be the first step

doron-st replied:
Will delete /1

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -326,41 +341,92 @@ process applyBQSR {
+}
 
-        """
+// Combine multiple VCFs from scattered HaplotypeCaller runs
```
add "Stage 11 -" prefix

doron-st replied:
Done

avigailmo commented on nf/maestro3/genotype_fastq.nf:
```diff
@@ -39,8 +41,13 @@ compression_level = 5
+input_bam_ch = input_bam_files.merge(input_bam_index_files).map { t -> [t[0].baseName, t[0], t[1]]}
+
 //Create files from input paths (localize)
-fastq_pairs = Channel.fromFilePairs("${params.fastq_dir}/*.{1,2}.fastq.gz", flat: true, checkIfExists: true)
```
why have you removed the `checkIfExists` validation? Don't we to fail if there isn't?

doron-st replied:
I wanted to allow fastq And/OR bam...
I agree that we should add validation that at least one input exists.

doron-st replied:
Done

doron-st commented:
> Generally looks good. Added some comments/ questions
> What happens in case of failure of one thread?
> Does it fail the whole stage immediately?

I assume you mean thread=process, stage=workflow?

If so, currently yes. If a process fails 3 times the entire workflow shuts down.
 I've looked it up and added a comment line to the config file showing how nextflow purposes to retry and eventualy IGNORE a repeating error. Not sure if we want to use this as the default.

doron-st commented on config/nextflow.aws.config:
```diff
@@ -3,6 +3,7 @@ executor.name = 'awsbatch'
   queue = 'GenomicsDefaultQueue-db299dfc1fe10b9'
   container = 'nrgene-docker.jfrog.io/nrgene-base:1.0'
+  //errorStrategy = { task.attempt <= 2 ? 'retry' : 'ignore' }
   errorStrategy = 'retry'
```
@avigailmo see here about the different possible errorStrategy.
